File: 13.roman-to-integer.py
#
# @lc app=leetcode id=13 lang=python3
#
# [13] Roman to Integer
#

# @lc code=start

DIGITS = { 'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000 }

class Solution:
    def romanToInt(self, s: str) -> int:
        n = 0
        for i, a in enumerate(s):
            va = DIGITS.get(a)
            vb = DIGITS.get(s[i + 1]) if i + 1 != len(s) else 0
            n += va if va >= vb else -va
        return n
# @lc code=end


print(Solution().romanToInt("XIX"))

File: 13.roman-to-integer.cpp
/*
 * @lc app=leetcode id=13 lang=cpp
 *
 * [13] Roman to Integer
 */

// @lc code=start
#include <string>
#include <unordered_map>

using namespace std;

const unordered_map<char, int> M = {
    {'I', 1},
    {'V', 5},
    {'X', 10},
    {'L', 50},
    {'C', 100},
    {'D', 500},
    {'M', 1000}
}; 

class Solution {
public:
    int romanToInt(string s)
    {
        int sum = 0;
        for (int i = 0; i < s.size(); ++i) {
            auto v = M.at(s[i]);
            if (i + 1 < s.size() && v < M.at(s[i+1])) {
                sum -= v;
            } else {
                sum += v;
            }
        }
        return sum;
    }
};
// @lc code=end

